### PR TITLE
RPC: Tolerate invalid rpcauth options

### PR DESCRIPTION
**Pull request by luke-jr**
By tolerating unknown extra rpcauth parameters (and ignoring the rpcauth), we can ensure a limited forward compatibility by not forcing users to downgrade their config file to switch back to older versions (perhaps temporarily).

Same as #20548, but without the additional effort to explain the situation at runtime.

### CODE REVIEW

File: src/httprpc.cpp [hunk 1]
```diff
@@ -260,7 +260,6 @@ static bool InitRPCAuthentication()
                 g_rpcauth.push_back(fields);
             } else {
                 LogPrintf("Invalid -rpcauth argument.\n");
```
Comment (MarcoFalke) [inline src/httprpc.cpp:262]:
```suggestion
                LogPrintf("Ignoring invalid -rpcauth argument.\n");
```

**Review by maflcko — APPROVED** [review]
review ACK 017b6f82de56db945eceee5f9f7ea5069ea541a9, but the release notes might need to be adjusted?

**Comment by promag** [comment]
Concept ACK, you have to update test/functional/rpc_users.py.

**Comment by luke-jr** [comment]
Release notes updated, debug msg changed, and tests updated.

**Review by promag — COMMENTED** [review]
Tested ACK 044df97886e1366ad995a03c0c4a0d62616f66a2.

**Comment by MarcoFalke** [comment]
review ACK 044df97886e1366ad995a03c0c4a0d62616f66a2

**Comment by laanwj** [comment]
I prefer failing here to be honest. But I guess my dislike for ignoring invalid options is well known.

**Comment by promag** [comment]
> I prefer failing here to be honest. But I guess my dislike for ignoring invalid options is well known.

In https://github.com/bitcoin/bitcoin/pull/20461#issuecomment-737365945 I've suggested a `-strict` (or similar) option.

**Comment by laanwj** [comment]
I mean, if the entire problem here is that users running older bitcoin core might run into this error when upgrading and be confused, I think the error needs to be clearer. All in all it will result in removing cruft from their configuration. Even more, they will likely also be alerted to mistakes they made.
(…why would one add entries on purpose just to be ignored?)

>  I've suggested a -strict (or similar) option.

Not sure this kind of meta-configuration makes sense. It just adds more code paths where it's better to make a principled decision. And I think bitcoin core should be strict and unambigious about input handling.

**Comment by jnewbery** [comment]
I totally agree with @laanwj here. We should fail on invalid configuration.

**Comment by luke-jr** [comment]
This still fails, just at runtime when it makes more sense to do so (and without disrupting other usage).

**Comment by harding** [comment]
> By tolerating unknown extra rpcauth parameters (and ignoring the rpcauth), we can ensure a limited forward compatibility by not forcing users to downgrade their config file to switch back to older versions (perhaps temporarily).

For the past couple releases, hasn't Bitcoin Core refused to start on unknown configuration settings?

    $ bitcoind -not-a-real-setting
    Error: Error parsing command line arguments: Invalid parameter -not-a-real-setting

Future releases, or spinoffs, are likely to support new configuration settings that old releases will not, so any users including those settings in their configuration files will need to edit those files anyway to use an old or baseline release.  Rejecting unknown parameters seems like a simple and logically consistent extension of this previous policy.

The policy of failing on unknown settings is quite beneficial at eliminating the annoyance of figuring out that your configuration is broken and then restarting the node---which is something that can take a long time (e.g. if you have a high db cache or are using an underpowered computer) and which inexperienced users might not know how to do (or might do wrong, e.g. sending SIGKILL rather than SIGTERM, which could result in a need to reindex).

Based on my own misadventures, wrongly configuring a node is a much more common problem than wanting to use a future/spinoff configuration file with an old/baseline node, so I think we should optimize for helping users fix misconfigurations as fast and painlessly as possible.

**Comment by MarcoFalke** [comment]
Closing for now due to controversy

**Comment by luke-jr** [comment]
>For the past couple releases, hasn't Bitcoin Core refused to start on unknown configuration settings?

On the command line only, not in config files where they would reasonably be expected.

**Comment by harding** [comment]
> On the command line only, not in config files

Confirmed.  That seems to me like an unnecessary half measure.  If it's planned to keep it that way, there may indeed be merit in a `-strict` option that also validates options in the configuration file.

**Comment by sipa** [comment]
For the configuration files we've been going in the same direction, e.g. `rpcport` causes failure outside of network sections.

**Comment by MarcoFalke** [comment]
Indeed, this is a bug that simply hasn't been fixed yet. See #15021

**Comment by sipa** [comment]
Right, I now remember. The reason rejecting unknown config options isn't done is because it would require all binaries to be aware of each other's options so they can be ignored.

**Comment by luke-jr** [comment]
That would just be gratuitously breaking things for no reason...

**Pull request closed without merging**